### PR TITLE
WIP: Datasets locking/v2

### PR DESCRIPTION
**Pull request by inashivb**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/7398

Previous PR: https://github.com/OISF/suricata/pull/12182

Changes since v1:
- v1 was incorrect, v2 is a WIP for moving the blocking call of reading from file to rust

### CODE REVIEW

File: rust/cbindgen.toml [hunk 1]
```diff
@@ -94,6 +95,7 @@ exclude = [
     "CLuaState",
     "DetectEngineState",
     "Flow",
+    "Dataset",
```
Comment (jasonish) [inline rust/cbindgen.toml:98]: Remove this, we need the opaque version of it as a forward declaration.

File: src/datasets.h [hunk 2]
```diff
@@ -19,6 +19,7 @@
 #define SURICATA_DATASETS_H
 
 #include "util-thash.h"
+#include "rust-bindings.h"
```
Comment (jasonish) [inline src/datasets.h:22]: Change this to `rust.h`.

File: rust/src/detect/datasets.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#[repr(C)]
+pub struct DataRepType {
+    pub value: u16,
+}
+
+// Extern fns operating on the opaque Dataset type above
+/// cbindgen:ignore
+extern {
+    pub fn DatasetAdd(set: &Dataset, data: *const u8, len: u32) -> i32;
+    pub fn DatasetAddwRep(set: &Dataset, data: *const u8, len: u32, rep: *const DataRepType) -> i32;
```
Comment (jasonish) [inline rust/src/detect/datasets.rs:43]: This over on the C sides need to be non-static.

File: rust/src/detect/datasets.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+use std::fs::File;
+use std::io::{self, BufRead};
+use std::path::Path;
+use std::ffi::{c_char, CStr};
+use base64::{Engine, engine::general_purpose::STANDARD};
+
+/// Opaque Dataset type defined in C
+#[derive(Copy, Clone)]
+pub enum Dataset {}
```
[the remark this answers is not in the document]
Comment (jasonish) [reply rust/src/detect/datasets.rs:30]: Builds with the few notes I made :)

[the remark this answers is not in the document]
Comment (inashivb) [reply rust/src/detect/datasets.rs:30]: Thank you so much! 🙇🏽‍♀️ You're awesome! 🤩

File: rust/src/detect/datasets.rs (hunk 3 above)
Comment (inashivb) [reply rust/src/detect/datasets.rs:43]: Had thought about this during development. Forgot later on 🤦🏽‍♀️

**Pull request closed without merging**